File: integrations/hyperspace-memory-ts/src/test_memory.ts
import { Memory } from "./index.js";

const HOST = process.env.HYPERSPACE_HOST || "localhost:50051";
const API_KEY = process.env.HYPERSPACE_API_KEY || "I_LOVE_HYPERSPACEDB";
const TEST_COLLECTION = `test_mem0_ts_${Date.now()}`;

async function runMemoryComprehensiveTestSuite() {
  console.log("==========================================================================");
  console.log("🧠 HYPERSPACEDB MEMORY SDK (TypeScript): COMPREHENSIVE E2E TEST SUITE");
  console.log("==========================================================================");
  console.log(`   Host: ${HOST}`);
  console.log(`   Test Collection: ${TEST_COLLECTION}\n`);

  const memory = new Memory({
    host: HOST,
    apiKey: API_KEY,
    collectionName: TEST_COLLECTION,
    quantization: "extreme"
  });

  let passed = 0;
  let failed = 0;

  function pass(label: string) {
    console.log(`   ✅ ${label}`);
    passed++;
  }

  function fail(label: string, detail?: string) {
    console.error(`   ❌ ${label}${detail ? `: ${detail}` : ""}`);
    failed++;
  }

  // ── STEP 1: Mem0 API Interface Compliance ──────────────────────────────────
  console.log("--------------------------------------------------------------------------");
  console.log("📋 STEP 1: Mem0 Interface Method Signatures Verification");
  console.log("--------------------------------------------------------------------------");
  const methods = ["add", "search", "getAll", "get", "update", "delete", "deleteAll", "reset", "history"];
  for (const m of methods) {
    if (typeof (memory as any)[m] === "function") {
      pass(`Method present: ${m}`);
    } else {
      fail(`Missing required Mem0 method: ${m}`);
    }
  }

  // ── STEP 2: Multi-Domain Memory Ingestion ──────────────────────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("💾 STEP 2: Ingesting Multi-Domain Memories (Medical, Finance, Infra)");
  console.log("--------------------------------------------------------------------------");

  // Alice: Medical Domain
  const addAlice1 = await memory.add(
    "Patient blood glucose is 14.2 mmol/L with rapid rising trend at 14:30.",
    { userId: "user_alice", agentId: "agent_medical", runId: "run_shift_01", metadata: { category: "glucose", severity: "high" } }
  );
  const memAlice1 = addAlice1.results[0].id;
  pass(`Added Alice Medical Memory 1: ${memAlice1}`);

  const addAlice2 = await memory.add(
    "Target glucose range before meals is 4.0 - 7.0 mmol/L.",
    { userId: "user_alice", agentId: "agent_medical", runId: "run_shift_01", metadata: { category: "target", severity: "normal" } }
  );
  const memAlice2 = addAlice2.results[0].id;
  pass(`Added Alice Medical Memory 2: ${memAlice2}`);

  // Bob: Finance Domain
  const addBob1 = await memory.add(
    "Invoice #9042 for SaaS subscription $99 paid via Stripe webhook at 15:00.",
    { userId: "user_bob", agentId: "agent_finance", runId: "run_billing_02", metadata: { category: "billing", amount: "99" } }
  );
  const memBob1 = addBob1.results[0].id;
  pass(`Added Bob Finance Memory 1: ${memBob1}`);

  // Charlie: DevOps Domain
  const addCharlie1 = await memory.add(
    "Kubernetes cluster upgraded to v1.30 on primary us-east-1 region.",
    { userId: "user_charlie", agentId: "agent_devops", runId: "run_infra_03", metadata: { category: "infrastructure" } }
  );
  const memCharlie1 = addCharlie1.results[0].id;
  pass(`Added Charlie DevOps Memory 1: ${memCharlie1}`);

  // ── STEP 3: Multi-User Isolation & Leakage Verification ───────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("🔍 STEP 3: Multi-User Isolation (0% Cross-User Leakage)");
  console.log("--------------------------------------------------------------------------");

  const searchAlice = await memory.search("glucose target insulin", { userId: "user_alice", limit: 5 });
  if (searchAlice.length > 0 && searchAlice.every(m => m.userId === "user_alice")) {
    pass(`Alice search returned ${searchAlice.length} memories, all scoped to user_alice`);
  } else {
    fail("Alice search failed or returned cross-user memories");
  }

  const hasLeakBob = searchAlice.some(m => m.userId === "user_bob" || m.memory.includes("Invoice #9042"));
  if (!hasLeakBob) {
    pass("Zero cross-user leakage: Bob finance memory absent from Alice search");
  } else {
    fail("CRITICAL: Bob finance memory leaked into Alice search results!");
  }

  const searchBob = await memory.search("Stripe invoice payment", { userId: "user_bob", limit: 5 });
  if (searchBob.length > 0 && searchBob.every(m => m.userId === "user_bob")) {
    pass(`Bob search returned ${searchBob.length} memories, all scoped to user_bob`);
  } else {
    fail("Bob search failed or returned cross-user memories");
  }

  const searchGhost = await memory.search("glucose invoice", { userId: "user_ghost", limit: 5 });
  if (searchGhost.length === 0) {
    pass("Non-existent user search correctly returned 0 memories");
  } else {
    fail("Non-existent user search returned unexpected results");
  }

  // ── STEP 4: Semantic Ranking Accuracy ─────────────────────────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("📊 STEP 4: Semantic Ranking & Relevance Accuracy (v5_Light Embeddings)");
  console.log("--------------------------------------------------------------------------");

  const rankResults = await memory.search("What is the targeted blood glucose level before meals?", { userId: "user_alice", limit: 5 });
  if (rankResults.length > 0) {
    const topHit = rankResults[0];
    if (topHit.memory.includes("Target glucose range")) {
      pass(`Semantic ranking #1 hit is exact target memory (Score: ${topHit.score})`);
    } else {
      fail(`Unexpected top hit: ${topHit.memory}`);
    }
  } else {
    fail("Semantic ranking search returned no results");
  }

  // ── STEP 5: Agent-ID & Run-ID Scoping ─────────────────────────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("🎯 STEP 5: Agent-ID and Run-ID Scoping Verification");
  console.log("--------------------------------------------------------------------------");

  const agentResults = await memory.search("blood glucose", { agentId: "agent_medical", limit: 5 });
  if (agentResults.length > 0 && agentResults.every(m => m.agentId === "agent_medical")) {
    pass(`Agent scoping verified: all ${agentResults.length} hits match agent_medical`);
  } else {
    fail("Agent scoping failed");
  }

  const runResults = await memory.search("glucose", { runId: "run_shift_01", limit: 5 });
  if (runResults.length > 0 && runResults.every(m => m.runId === "run_shift_01")) {
    pass(`Run scoping verified: all ${runResults.length} hits match run_shift_01`);
  } else {
    fail("Run scoping failed");
  }

  // ── STEP 6: Update Memory & History Lineage ───────────────────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("✏️  STEP 6: Memory Update & Historical Revisions (update, history)");
  console.log("--------------------------------------------------------------------------");

  const updateRes = await memory.update(memAlice1, "Patient blood glucose normalized to 6.2 mmol/L after 4 units Humalog.");
  if (updateRes.message.includes("updated successfully")) {
    pass(`Updated Alice Memory ${memAlice1}: ${updateRes.message}`);
  } else {
    fail("Memory update failed", JSON.stringify(updateRes));
  }

  const hist = await memory.history(memAlice1);
  if (Array.isArray(hist)) {
    pass(`Retrieved history for ${memAlice1}: ${hist.length} record(s)`);
  } else {
    fail("History retrieval returned non-array");
  }

  // ── STEP 7: Granular Retrieval (get, getAll) ──────────────────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("📜 STEP 7: Granular Retrieval (get, getAll)");
  console.log("--------------------------------------------------------------------------");

  const aliceAll = await memory.getAll({ userId: "user_alice" });
  pass(`getAll for user_alice returned ${aliceAll.length} items`);

  const bobItem = await memory.get(memBob1);
  if (bobItem && bobItem.userId === "user_bob") {
    pass(`get(${memBob1}) retrieved Bob's memory: "${bobItem.memory.substring(0, 40)}..."`);
  } else {
    fail(`Failed to retrieve memory by ID ${memBob1}`);
  }

  // ── STEP 8: Granular Delete & Isolation ───────────────────────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("🗑️  STEP 8: Granular Delete & Unaffected Isolation (delete)");
  console.log("--------------------------------------------------------------------------");

  const delRes = await memory.delete(memBob1);
  pass(`Deleted Bob memory ${memBob1}: ${delRes.message}`);

  const aliceAfterDel = await memory.search("glucose", { userId: "user_alice", limit: 5 });
  if (aliceAfterDel.length > 0) {
    pass(`Alice memories intact after Bob deletion (${aliceAfterDel.length} memories retained)`);
  } else {
    fail("Alice memories were unexpectedly lost after Bob deletion");
  }

  // ── STEP 9: Collection Reset & Tear Down ──────────────────────────────────
  console.log("\n--------------------------------------------------------------------------");
  console.log("🧹 STEP 9: Reset Memory Collection & Tear Down (reset)");
  console.log("--------------------------------------------------------------------------");

  const resetRes = await memory.reset();
  pass(`Reset collection: ${resetRes.message}`);

  // ── Final Summary ─────────────────────────────────────────────────────────
  console.log("\n==========================================================================");
  console.log("📊 TYPESCRIPT MEMORY SDK TEST RESULTS");
  console.log("==========================================================================");
  console.log(`   ✅ PASSED: ${passed}`);
  console.log(`   ❌ FAILED: ${failed}`);
  console.log(`   Total:    ${passed + failed}`);
  console.log("==========================================================================");

  if (failed > 0) {
    throw new Error(`${failed} test(s) failed in TypeScript Memory SDK suite`);
  }
  console.log("🎉 ALL TYPESCRIPT COMPREHENSIVE MEMORY TESTS PASSED!\n");
}

runMemoryComprehensiveTestSuite().catch(err => {
  console.error("❌ Test failed:", err);
  process.exit(1);
});

File: integrations/hyperspace-memory-py/tests/test_memory.py
import os
import time
import unittest
from hyperspace_memory import Memory

HOST = os.environ.get("HYPERSPACE_HOST", "localhost:50051")
API_KEY = os.environ.get("HYPERSPACE_API_KEY", "I_LOVE_HYPERSPACEDB")
COLLECTION = f"test_mem0_py_{int(time.time())}"

class TestHyperspaceMemoryComprehensive(unittest.TestCase):
    """
    Comprehensive test suite for hyperspace-memory-py (Mem0 Drop-in replacement).
    Covers:
      1. Full interface compliance (all 9 Mem0 methods)
      2. Multi-domain memory storage (Medical, Finance, Infrastructure)
      3. Cross-user isolation (0% data leakage)
      4. Semantic ranking & score validity with real v5_Light embeddings
      5. Agent and Run-ID scoping & filtering
      6. Memory updating & historical lineage
      7. Granular retrieval (get, get_all)
      8. Deletion & collection reset
    """

    @classmethod
    def setUpClass(cls):
        print("\n" + "=" * 76)
        print("🧠 HYPERSPACE-MEMORY-PY: ADVANCED COMPREHENSIVE TEST SUITE")
        print("=" * 76)
        print(f"   Host: {HOST}")
        print(f"   Collection: {COLLECTION}\n")

        cls.memory = Memory(config={
            "host": HOST,
            "api_key": API_KEY,
            "collection_name": COLLECTION,
            "quantization": "extreme"
        })

    def test_01_interface_structure(self):
        """Step 1: Verify all 9 Mem0 API methods exist on Memory instance."""
        required_methods = [
            "add", "search", "get_all", "get", "update",
            "delete", "delete_all", "reset", "history"
        ]
        for method in required_methods:
            self.assertTrue(
                hasattr(self.memory, method),
                f"Missing required Mem0 compatibility method: {method}"
            )
        print(f"   ✓ Step 1: All {len(required_methods)} Mem0 compatibility methods present")

    def test_02_multi_domain_storage_and_user_isolation(self):
        """Step 2 & 3: Store memories across domains and verify 0% cross-user leakage."""
        # Domain 1: Medical (User Alice)
        res1 = self.memory.add(
            "Patient blood glucose is 14.2 mmol/L with rapid rising trend at 14:30.",
            user_id="user_alice",
            agent_id="agent_medical",
            run_id="run_shift_01",
            metadata={"category": "glucose", "severity": "high"}
        )
        self.assertIn("results", res1)
        self.__class__.mem_alice_1 = res1["results"][0]["id"]

        res2 = self.memory.add(
            "Target glucose range before meals is 4.0 - 7.0 mmol/L.",
            user_id="user_alice",
            agent_id="agent_medical",
            run_id="run_shift_01",
            metadata={"category": "target", "severity": "normal"}
        )
        self.__class__.mem_alice_2 = res2["results"][0]["id"]

        # Domain 2: Finance (User Bob)
        res3 = self.memory.add(
            "Invoice #9042 for SaaS subscription $99 paid via Stripe webhook at 15:00.",
            user_id="user_bob",
            agent_id="agent_finance",
            run_id="run_billing_02",
            metadata={"category": "billing", "amount": "99"}
        )
        self.__class__.mem_bob_1 = res3["results"][0]["id"]

        # Domain 3: Infrastructure (User Charlie)
        res4 = self.memory.add(
            "Kubernetes cluster upgraded to v1.30 on primary us-east-1 region.",
            user_id="user_charlie",
            agent_id="agent_devops",
            run_id="run_infra_03",
            metadata={"category": "infrastructure"}
        )
        self.__class__.mem_charlie_1 = res4["results"][0]["id"]

        print("   ✓ Step 2: Ingested 4 multi-domain memories across 3 distinct users")

        # Isolation check: Query as Alice
        search_alice = self.memory.search("glucose target insulin", user_id="user_alice", limit=5)
        self.assertGreater(len(search_alice), 0, "Alice search returned 0 results")
        for item in search_alice:
            self.assertEqual(item.get("user_id"), "user_alice", "Cross-user memory leak detected!")
            self.assertNotIn("Invoice #9042", item.get("memory", ""))
            self.assertNotIn("Kubernetes", item.get("memory", ""))

        # Isolation check: Query as Bob
        search_bob = self.memory.search("Stripe invoice payment", user_id="user_bob", limit=5)
        self.assertGreater(len(search_bob), 0, "Bob search returned 0 results")
        for item in search_bob:
            self.assertEqual(item.get("user_id"), "user_bob", "Cross-user memory leak detected!")
            self.assertNotIn("glucose", item.get("memory", "").lower())

        # Isolation check: Unknown user
        search_unknown = self.memory.search("glucose invoice", user_id="user_ghost", limit=5)
        self.assertEqual(len(search_unknown), 0, "Non-existent user must have 0 memories")

        print("   ✓ Step 3: Multi-user isolation VERIFIED (0% cross-user leakage)")

    def test_03_semantic_ranking_accuracy(self):
        """Step 4: Verify semantic ranking relevance and score validity."""
        # Query specific to target range
        query = "What is the recommended blood sugar target before meals?"
        results = self.memory.search(query, user_id="user_alice", limit=5)
        self.assertGreaterEqual(len(results), 1)

        top_hit = results[0]
        self.assertIn("Target glucose range", top_hit["memory"])
        self.assertGreater(top_hit["score"], 0.0, "Score must be positive")

        # Verify score descending order
        if len(results) > 1:
            for i in range(len(results) - 1):
                self.assertGreaterEqual(
                    results[i]["score"],
                    results[i + 1]["score"],
                    "Search results must be sorted in descending score order"
                )
        print(f"   ✓ Step 4: Semantic ranking VERIFIED (Top score: {top_hit['score']:.4f})")

    def test_04_agent_and_run_filtering(self):
        """Step 5: Verify agent_id and run_id partition isolation."""
        # Filter by agent_id
        med_results = self.memory.search("blood glucose", agent_id="agent_medical", limit=5)
        self.assertGreater(len(med_results), 0)
        for r in med_results:
            self.assertEqual(r.get("agent_id"), "agent_medical")

        # Filter by run_id
        shift_results = self.memory.search("glucose", run_id="run_shift_01", limit=5)
        self.assertGreater(len(shift_results), 0)
        for r in shift_results:
            self.assertEqual(r.get("run_id"), "run_shift_01")

        print("   ✓ Step 5: Agent-ID and Run-ID scoping VERIFIED")

    def test_05_update_and_history(self):
        """Step 6: Verify memory updating and revision history."""
        mem_id = self.__class__.mem_alice_1
        updated_text = "Patient blood glucose normalized to 6.2 mmol/L after 4 units Humalog."
        
        upd_res = self.memory.update(mem_id, updated_text)
        self.assertIn("updated successfully", upd_res.get("message", ""))
        self.assertEqual(upd_res.get("data"), updated_text)

        # Verify updated memory is retrievable
        mem = self.memory.get(mem_id)
        if mem:
            self.assertIn("6.2 mmol/L", mem.get("memory", ""))

        # Verify history
        hist = self.memory.history(mem_id)
        self.assertIsInstance(hist, list)

        print("   ✓ Step 6: Memory update and history revision VERIFIED")

    def test_06_get_and_get_all(self):
        """Step 7: Verify single get and user-scoped get_all."""
        alice_memories = self.memory.get_all(user_id="user_alice")
        self.assertGreaterEqual(len(alice_memories), 1)
        for m in alice_memories:
            self.assertEqual(m.get("user_id"), "user_alice")

        bob_mem = self.memory.get(self.__class__.mem_bob_1)
        if bob_mem:
            self.assertEqual(bob_mem.get("user_id"), "user_bob")

        print(f"   ✓ Step 7: Retrieval (get, get_all) VERIFIED ({len(alice_memories)} active for Alice)")

    def test_07_granular_delete(self):
        """Step 8: Verify targeted delete removes only specified item."""
        bob_id = self.__class__.mem_bob_1
        del_res = self.memory.delete(bob_id)
        self.assertIn("deleted successfully", del_res.get("message", ""))

        # Verify Alice memories still intact
        alice_after = self.memory.search("glucose", user_id="user_alice", limit=5)
        self.assertGreater(len(alice_after), 0, "Alice memories must not be affected by Bob's deletion")

        print("   ✓ Step 8: Granular deletion VERIFIED")

    def test_08_reset_cleanup(self):
        """Step 9: Verify complete collection reset and purge."""
        reset_res = self.memory.reset()
        self.assertIn("reset successfully", reset_res.get("message", ""))
        print("   ✓ Step 9: Collection reset & tear down VERIFIED")
        print("\n" + "=" * 76)
        print("🎉 ALL 9 PYTHON HYPERSPACE-MEMORY COMPREHENSIVE TESTS PASSED!")
        print("=" * 76 + "\n")

if __name__ == "__main__":
    unittest.main()
